openmdao/components/subproblem_comp.py

```diff
@@ -0,0 +1,310 @@
+        If _UNDEFINED, the OPENMDAO_REPORTS variable is used. Defaults to _UNDEFINED.
+        If given, reports overrides OPENMDAO_REPORTS. If boolean, enable/disable all reports.
+        Since none is acceptable in the environment variable, a value of reports=None
+        is equivalent to reports=False. Otherwise, reports may be a sequence of
+        strings giving the names of the reports to run.
+    prob_options : dict or None
+        Remaining named args for problem that are converted to options.
+    **kwargs : named args
+        All remaining named args that become options for `SubproblemComp`.
+
+    Attributes
+    ----------
+    prob_args : dict
+        Extra arguments to be passed to the problem instantiation.
+    model : <System>
+        The system being analyzed in subproblem.
+    model_input_names : list of str or tuple
+        List of inputs requested by user to be used as inputs in the
+        subproblem's system.
+    model_output_names : list of str or tuple
+        List of outputs requested by user to be used as outputs in the
+        subproblem's system.
+    """
+
+    def __init__(self, model, inputs=None, outputs=None, driver=None, comm=None,
+                 name=None, reports=_UNDEFINED, prob_options=None, **kwargs):
+        """
+        Initialize all attributes.
+        """
+        # check for driver and issue warning about its current use
+        # in subproblem
+        if driver is not None:
+            issue_warning('Driver results may not be accurate if'
+                          ' derivatives are needed. Set driver to'
+                          ' None if your subproblem isn\'t reliant on'
+                          ' a driver.')
+
+        # make `prob_options` empty dict to be passed as **options to problem
+        # instantiation
+        if prob_options is None:
+            prob_options = {}
+
+        # call base class to set kwargs
+        super().__init__(**kwargs)
+
+        # store inputs and outputs in options
+        self.options.declare('inputs', {}, types=dict,
+                             desc='Subproblem Component inputs')
+        self.options.declare('outputs', {}, types=dict,
+                             desc='Subproblem Component outputs')
+
+        # set other variables necessary for subproblem
+
+        self.prob_args = {'driver': driver,
+                          'comm': comm,
+                          'name': name,
+                          'reports': reports}
+
+        self.prob_args.update(prob_options)
+
+        self.model = model
+        self.model_input_names = inputs
+        self.model_output_names = outputs
+
+    def setup(self):
+        """
+        Perform some final setup and checks.
+        """
+        p = self._subprob = Problem(**self.prob_args)
+        p.model.add_subsystem('subsys', self.model, promotes=['*'])
+
+        p.setup(force_alloc_complex=self._problem_meta['force_alloc_complex'])
+        p.final_setup()
```

Comment:
Now that I'm experimenting with this in dymos, I think that the first 4 lines of code of `setup` should be moved to `__init__`.  I have a use-case where I want to be able to interrogate the subproblem for its outputs and unconnected inputs during my setup process, which requires that the subproblem's setup process be done already.

Because I don't necessarily know what inputs and outputs are to be transferred to/from the subproblem, we should also implement `add_input` and `add_output` methods specific to this component that do something along the lines of:

```
def add_subproblem_input(self, path, name=None):
    self.model_input_names.append((path, name))
   
def add_subproblem_output(self, path, name=None):
    self.model_output_names.append((path, name)) 
```

Reply:
@naylor-b I would appreciate your thoughts on this.

Reply:
We should probably add some error checking with these as well so that add_input/add_output can't be called after a certain point in setup.  Otherwise the calls will basically be ignored.  We should also add a test if we don't have one already that calls setup multiple times when a subprob comp is present just to make sure we don't have any weird setup issues.

Reply:
Implemented error checking in `add_input` and `add_output` to make sure they don't occur after configure. Also implemented test `test_multiple_setups` to ensure SubproblemComp doesn't have any weird setup issues.